Add dotconfig property for c_cpp_properties.json

microhobby opened the pull request:
Some bare metal systems, RTOS and Linux use a generated .config file
with defines needed to build the system and switch features. This patch
add the `dotconfig` property that will store the path for read a
.config file and will concat the defines from file to the extension
defines.

Signed-off-by: Matheus Castello <matheus@castello.eng.br>

microhobby committed:
Add dotConfig property for c_cpp_properties.json

Kconfig system generate a .config file with all the defines, CONFIG
options selected, to build a project. Examples of projects that use
Kconfig system is Linux Kernel and NuttX RTOS. This patch add the
`dotConfig` property that will store the path for read a .config file
and will concat the defines from file to the extension defines.

Signed-off-by: Matheus Castello <matheus@castello.eng.br>

sean-mcmanus commented on Extension/src/LanguageServer/configurations.ts:
```diff
@@ -764,6 +781,10 @@ export class CppProperties {
             configuration.defines = this.updateConfigurationStringArray(configuration.defines, settings.defaultDefines, env);
+
+            configuration.dotconfig = this.updateConfigurationString(configuration.dotconfig, settings.defaultDotconfig, env);
+            configuration.defines = configuration.defines?.concat(this.getDotconfigDefines(configuration.dotconfig as string));
```
Can configuration.defines be null or undefined? If so, I don't think this uses the .config defines.

microhobby replied:
hmm, yeah the linter complained if I don't use the `?` here, you are right, let me check this

bobbrow commented on Extension/c_cpp_properties.schema.json:
```diff
@@ -82,6 +82,10 @@
                         "type": "string",
                         "pattern": "^\\d{2}\\.\\d{1}\\.\\d{5}\\.\\d{1}$|^8\\.1$"
                     },
+                    "dotconfig": {
```
Please update the name to `dotConfig` to match our property naming style.

Colengms commented on Extension/c_cpp_properties.schema.json:
```diff
@@ -82,6 +82,10 @@
                         "pattern": "^\\d{2}\\.\\d{1}\\.\\d{5}\\.\\d{1}$|^8\\.1$"
                     },
+                    "dotconfig": {
+                        "description": "A path to .config file",
```
Hi @microhobby .  Could you provide more information here about `.config` files in this description field?  Is the convention of using `.config` files specific to Linux kernel development?  Perhaps refer to that in the description, for users who are unfamiliar with the file and may not otherwise understand what this setting would do.  Also, since the title of the PR contains "WIP", I assume it's not ready to be signed off on.  When you are ready for us to sign off, could you remove that from the title?

Colengms commented on Extension/src/LanguageServer/configurations.ts:
```diff
@@ -746,6 +747,22 @@ export class CppProperties {
+                return configContent.filter(i => !i.startsWith("#") && i !== "");
+            }
+        } catch (err) {
+            const failedToResolve: string = localize("cannot.resolve.dotconfig.path", 'Invalid input, cannot resolve .config path');
```
Since only the readFileSync call would throw, could you wrap only that with the try/catch?  Also, could you remove ".config" from the localizable portion of the string (to avoid having it translated)?  i.e.: `localize("cannot.resolve.dotconfig.path", "Cannot resolve {0} path: {1}", ".config");`  Also, consider logging an error to our output channel using  `getOutputChannelLogger().appendLine()` instead of display an error dialog.  (Users seem to dislike error dialogs).

microhobby replied:
@Colengms I updated it and also added another commit to validate the path to check if the path is valid and if the path is a valid file, adding warning diagnostics to point it to the user during the file edition. Let me know what you think.

microhobby committed:
Validate dotConfig path

Validate if the path exists and if the path is a valida file. For these
add a diagnostic warning to show a squiggle during the file edition.

Signed-off-by: Matheus Castello <matheus@castello.eng.br>

sean-mcmanus commented on Extension/c_cpp_properties.schema.json:
```diff
@@ -82,6 +82,10 @@
                         "pattern": "^\\d{2}\\.\\d{1}\\.\\d{5}\\.\\d{1}$|^8\\.1$"
                     },
+                    "dotConfig": {
+                        "description": "A path to a .config file created by Kconfig system. Kconfig system generate a file with all the defines, CONFIG options selected, to build a project. Examples of projects that use Kconfig system is Linux Kernel and NuttX RTOS.",
```
generate->generates

sean-mcmanus replied:
I don't understand how this ", CONFIG options selected," fits into the sentence -- should the commas be removed and the 1st comma replaced by an "and" or is there another intended meaning?

sean-mcmanus replied:
is->are

sean-mcmanus commented on Extension/c_cpp_properties.schema.json:
```diff
@@ -82,6 +82,10 @@
                         "pattern": "^\\d{2}\\.\\d{1}\\.\\d{5}\\.\\d{1}$|^8\\.1$"
                     },
+                    "dotConfig": {
+                        "description": "A path to a .config file created by Kconfig system. Kconfig system generate a file with all the defines, CONFIG options selected, to build a project. Examples of projects that use Kconfig system is Linux Kernel and NuttX RTOS.",
```
"Linux kernel"->"the Linux kernel"

microhobby replied:
@sean-mcmanus Proposed changes applied, thanks for the review

sean-mcmanus commented on Extension/ui/settings.ts:
```diff
@@ -260,6 +261,7 @@ class SettingsApp {
 
             (<HTMLInputElement>document.getElementById(elementId.intelliSenseMode)).value = config.intelliSenseMode ? config.intelliSenseMode : "${default}";
             (<HTMLInputElement>document.getElementById(elementId.includePath)).value = joinEntries(config.includePath);
+            (<HTMLInputElement>document.getElementById(elementId.dotConfig)).value = joinEntries(config.dotConfig);
```
Can you move this to the end of the settings (to the Advanced section)? Same for the case above with the elementId.

microhobby replied:
@sean-mcmanus proposed changes applied, thanks for the review.

sean-mcmanus reviewed (changes requested):
It seems like the package.json is missing the "default.dotConfig" setting and the package.nls.json is missing a description for that. I have a bunch of settings added in my branch at https://github.com/microsoft/vscode-cpptools/pull/7816/files if you want an example, but you don't need to use markdownDescription since that is only used in that branch and I can update it later.

microhobby committed:
Add dotConfig to package.json/package.nls.json

Signed-off-by: Matheus Castello <matheus@castello.eng.br>

michelleangela reviewed (changes requested):
This property will also need to be added in the .html file of the IntelliSense Configuration UI.

See https://github.com/microsoft/vscode-cpptools/blob/main/Extension/ui/settings.html and the UI element for "Compile commands" property for example of creating a field for one file path entry.

Place UI element for dotConfig after "Forced include" in the UI. Use title "Dot Config" for the new property.

Also add this new property to https://github.com/microsoft/vscode-cpptools/blob/main/Extension/src/LanguageServer/settingsPanel.ts so that the UI will be updated with values from c_cpp_properties.json

See examples for `elementId.compileCommands` in settingsPanel.ts.

https://github.com/microsoft/vscode-cpptools/blob/5dcb3081e502f84185b71f773f3d106b5d27f63c/Extension/src/LanguageServer/settingsPanel.ts#L41
https://github.com/microsoft/vscode-cpptools/blob/5dcb3081e502f84185b71f773f3d106b5d27f63c/Extension/src/LanguageServer/settingsPanel.ts#L325

michelleangela commented on Extension/ui/settings.ts:
```diff
@@ -270,6 +271,7 @@ class SettingsApp {
             (<HTMLInputElement>document.getElementById(elementId.compileCommands)).value = config.compileCommands ? config.compileCommands : "";
             (<HTMLInputElement>document.getElementById(elementId.configurationProvider)).value = config.configurationProvider ? config.configurationProvider : "";
             (<HTMLInputElement>document.getElementById(elementId.forcedInclude)).value = joinEntries(config.forcedInclude);
+            (<HTMLInputElement>document.getElementById(elementId.dotConfig)).value = joinEntries(config.dotConfig);
```
Since `config.dotConfig` will only have one file path, this does not need `joinEntries`.

For setting elements with one value:
```
(<HTMLInputElement>document.getElementById(elementId.dotConfig)).value = config.dotConfig ? config.dotConfig : "";
```

microhobby committed:
dotConfig use markDownDescription instead description

Signed-off-by: Matheus Castello <matheus@castello.eng.br>

microhobby committed:
Fix unnecessary use of joinEntries for elementId.dotConfig set value

Signed-off-by: Matheus Castello <matheus@castello.eng.br>

microhobby committed:
Add dotConfig to the Configuration UI

Signed-off-by: Matheus Castello <matheus@castello.eng.br>

microhobby committed:
Need to define error handler type before use it

Signed-off-by: Matheus Castello <matheus@castello.eng.br>

microhobby committed:
Show validate error message for dotConfig on Configuration UI

Signed-off-by: Matheus Castello <matheus@castello.eng.br>
